src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,296 @@
+/-
+Copyright (c) 2021 Hunter Monroe. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hunter Monroe, Kyle Miller, Alena Gusakov
+-/
+import combinatorics.simple_graph.basic
+import data.set.finite
+
+/-!
+# Subgraphs of a simple graph
+
+A subgraph of a simple graph consists of subsets of the graph's vertices and edges such that the
+endpoints of each edge are present in the vertex subset.  The edge subset is formalized as a
+sub-relation of the adjacency relation of the simple graph.
+
+## Main definitions
+
+* `subgraph G` is the type of subgraphs of a `G : simple_graph`
+
+* `subgraph.neighbor_set`, `subgraph.incidence_set`, and `subgraph.degree` are like their
+  `simple_graph` counterparts, but they refer to vertices from `G` to avoid subtype coercions.
+
+* `subgraph.coe` is the coercion from a `G' : subgraph G` to a `simple_graph G'.verts`.
+  (This cannot be a `has_coe` instance due to universe variable inference issues.)
+
+* A `bounded_lattice (subgraph G)` instance.
+
+* Graph homomorphisms from a subgraph to a graph (`subgraph.map_top`) and between subgraphs
+  (`subgraph.map`).
+
+## Todo
+
+* Images of graph homomorphisms as subgraphs.
+
+-/
+
+universe u
+
+namespace simple_graph
+
+/-- A subgraph of a `simple_graph` is a subset of vertices along with a restriction of the adjacency
+relation that is symmetric and is supported by the vertex subset.  They also form a bounded lattice.
+
+Thinking of `V → V → Prop` as `set (V × V)`, a set of darts, then `subgraph.adj_sub` is that the
+darts of a subgraph are a subset of the darts of `G`. -/
+@[ext]
+structure subgraph {V : Type u} (G : simple_graph V) :=
+(verts : set V)
+(adj : V → V → Prop)
+(adj_sub : ∀ ⦃v w : V⦄, adj v w → G.adj v w)
+(edge_vert : ∀ ⦃v w : V⦄, adj v w → v ∈ verts)
+(sym : symmetric adj . obviously)
+
+namespace subgraph
+
+variables {V : Type u} {G : simple_graph V}
+
+lemma adj_comm (G' : subgraph G) {v w : V} : G'.adj v w ↔ G'.adj w v :=
```

Comment:
It varies between being explicit and implicit, so I thought it would be less pain doing it this way.

For an easy-to-type-out argument like this, it seems like there's no harm including it.

Reply:
does the `(G)` and `{G}` syntax not work?

Reply:
That creates a new variable if used in a definition, rather than changing the explicitness of the existing one

Reply:
Ahh, I see

Reply:
Do we want `G'` explicit here anyway?

Reply:
@YaelDillies A nice thing having it be explicit is that it lets you do `rw [G'.adj_comm]` if you have a few subgraphs floating around.

I missed that the `simple_graph` version has the `v` and `w` arguments explicit, too -- I'll change this to match.

Reply:
I'd rather make the `simple_graph` ones implicit. It's usually better for equivalences.

Reply:
If this should change, this should be another PR -- best be consistent for now.  (As a counterpoint, `mul_comm` takes explicit arguments.)